mlrun/datastore/redis.py

```diff
@@ -0,0 +1,136 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+import redis
+import redis.cluster
+from storey.redis_driver import RedisType
+
+import mlrun
+
+from .base import DataStore
+
+
+class RedisStore(DataStore):
+    """
+    Partial implementation of DataStore over the Redis KV store
+
+    - no support for filesystem
+    - key and value sizes are limited to 512MB
+    """
+
+    def __init__(
+        self,
+        parent,
+        schema,
+        name,
+        endpoint="",
+    ):
+        super().__init__(parent, name, schema, endpoint)
+        self.headers = None
+
+        self.endpoint = self.endpoint or mlrun.mlconf.redis_url
+
+        if self.endpoint.startswith("rediss://"):
+            self.endpoint = self.endpoint[len("rediss://") :]
+            self.secure = True
+        elif self.endpoint.startswith("redis://"):
+            self.endpoint = self.endpoint[len("redis://") :]
+            self.secure = False
+        else:
+            raise NotImplementedError(f"invalid endpoint: {endpoint}")
+
+        self._redis_url = f"{schema}://{self.endpoint}"
+
+        self._redis = None
+
+        if os.environ.get("MLRUN_REDIS_TYPE") == "cluster":
+            self._redis_type = RedisType.CLUSTER
+        else:
+            self._redis_type = RedisType.STANDALONE
+
+    @property
+    def redis(self):
+        if self._redis is None:
+            if self._redis_type is RedisType.STANDALONE:
+                self._redis = redis.Redis.from_url(
+                    self._redis_url, decode_responses=True
+                )
+            else:
+                self._redis = redis.cluster.RedisCluster.from_url(
+                    self._redis_url, decode_response=True
+                )
+        return self._redis
+
+    def get_filesystem(self, silent):
+        return None  # no support for fsspec
+
+    def supports_isdir(self):
+        return False
+
+    def upload(self, key, src_path):
+        with open(src_path, "rb") as f:
+            while True:
+                data = f.read(1000 * 1000)
+                if not data:
+                    break
+                self.redis.append(key, data)
+
+    def get(self, key, size=-1, offset=0):
+        if size == 0:
+            raise mlrun.errors.MLRunInvalidArgumentError(
+                "size argument should not be zero"
+            )
+        if offset:
+            end_offset = size if size == -1 else offset + size - 1
+            return self.redis.getrange(key, offset, end_offset)
+        else:
+            return self.redis.get(key)
```

Comment:
Now I cannot call `get(key, size=X)` to get only the first X bytes. This will return the entire object...
Note also that if I use negative size that is not -1, for example if I expect to use `size=-2` to call `getrange` for everything but the last 2 chars, then it will also not work (will return nothing).
I think we're having trouble with this function since the signature uses start/size while Redis' `getrange` uses start/end. We're trying to force the `getrange` semantics here and it causes issues. Right now, no other datastore in MLRun used negative size like this, and I think we should just stick to it. Keep the logic simple:
- `size` should be by default `None` which indicates "everything to the end of the object" (`getrange` with -1 as `end`)
- Can assert on `size=0` like already done and even on `size<0` (or not, in which case nothing will be returned)
- For a positive size, calculate the `end_offset` like already done

If the `getrange` functionality is desired, maybe add a separate function which gets `start` and `end` and passed them to `getrange` and that's it.

Reply:
fixed